### ISSUE TITLE
LocalSymbol and BoundLocal types should match in error scenarios

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#34158]
Consider this code:

```C#
class Program
{
    static void F()
    {
        var x = new[] { x };
    }
}
```
The `BoundLocal` representing the reference to `x` in the array initializer has a `Type` of `ErrorType var`, but the `BoundLocal` itself has a `Type` of `var[]`. These should match.

**Comment by gafter** [comment]
I disagree with the issue as described.  The reference to `x` in the initializer is indeed an error as seen from the initial binder, and should be represented by an error type.  There is no type that could possibly be computed in the initial binder for the type of `x` that would be correct other than an error type.  That's because the type for `x` in the initializer must be computed before the type for the local variable can be set.  Whatever type we use for the reference to `x` in the array initializer, the type of `x` in the declaration is an array of that type.

The types of the bound node in initial binding and the type computed in the nullable walker should agree modulo nullability for all bound nodes (and not be relaxed to accept error vs non-error).  If the initial binder computes an error type, that is the type that should result from the nullable walker.

See also https://github.com/dotnet/roslyn/issues/34993.

**Comment by 333fred** [comment]
@gafter that's fine, but if that's the resolution we want then we need to update the nullable walker, as VisitLocal assumes that the boundnode and the symbol for the local have the same type currently.
